### DISCUSSION TITLE
Should the `data` property on eden treaty post response be typed?

### DISCUSSION THREAD

**Discussion by hill** [elysiajs/elysia#886] [Q&A]
I'm very new to Elysia and am trying to understand its capabilities.

When reading [this doc](https://elysiajs.com/eden/treaty/response) about the Eden treaty response, I understand that the type of the `data` field should be typed - as suggested in the code snippet on that page:

```ts
const submit = async (name: string) => {
    const { data, error } = await api.user.post({
        name
    })

    // type: string | null
    console.log(data)
```

However when I try to do this with a monorepo setup (A client package and server package, with the tsconfig.json in client referencing the server directory), the typehint has `data` as `any`:

<img width="396" alt="Screenshot 2024-10-20 at 6 32 17 PM" src="https://github.com/user-attachments/assets/c30f8890-a53e-4cf4-a7a6-8bdd75247ba2">

I noticed in the [overview docs](https://elysiajs.com/eden/overview.html#eden-treaty-recommended), that the post data response (`nendoroid`) is typed as `any` on hover:

<img width="740" alt="Screenshot 2024-10-20 at 6 33 41 PM" src="https://github.com/user-attachments/assets/2669b072-daf1-40f6-b029-ff77d29562c7">

Is it intended that the `post()` method does not have `data` type inferred from the server response?

Sorry If I missed documentation outlining this - would love to be pointed to it.

For reference my server code looks like:

```ts
import { cors } from "@elysiajs/cors";
import swagger from "@elysiajs/swagger";
import { Elysia, t } from "elysia";
import getBookcover from "./bookcover";

export const app = new Elysia()
	.use(cors())
	.use(swagger())
	.get("/", () => "Hello Bookshelf")
	.post(
		"/book",
		({ body }) => {
			const details = getBookcover({ title: body.title });
			return details;
		},
		{
			body: t.Object({
				title: t.String(),
			}),
			response: t.Object({
				cover: t.String(),
				isbn: t.String(),
			}),
		},
	)
	.listen(3000);
```

and my treaty:

```ts
import { treaty } from "@elysiajs/eden";
import type { App } from "@library/server";

export const api = treaty<App>("localhost:3000");
```

My client call is:

```ts
const response = await api.book.post({ title });

response.data // (typed as any)
```

In client dependencies,
`"@elysiajs/eden": "^1.1.3",`

and on server dependencies:
`"elysia": "^1.1.3"`

Thanks for your help!

**Comment by theopomies** [comment]
@hill try adding @types/bun to your devDependencies on the client, this fixed it for me after @SaltyAom suggested it

**Comment by Manuelbaun** [comment]
I also have this issue and adding @types/bun did not solve the problem.

for performance reason, I compile the app types into index.d.ts and import it in my frontend.

when using streams, data is of type any. but my compiled types I see that response 200 is of type AsyncGenerator, see images.
When using regular post, get etc, data is correctly inferred


![image](https://github.com/user-attachments/assets/de8ff368-3b06-4683-8ff8-08a71d66f4fc)
![image](https://github.com/user-attachments/assets/04d26445-fdca-47e5-9550-9577e85d2d55)

**Reply by Manuelbaun** [reply]
I found my error, here is the abbreviated endpoint code:

I return the error instead of throwing.And when I throw the error, I get the correct type! 


```ts
    .post(
        "/chat/:id/stream",
        async function* ({ body, params, error }) {
            const { reader, chatId, status, statusText, data } = await featureAIClient.chatStream();

            if (status !== 200) {
                return error(status, data || statusText);  // ^^ returning error instead of throwing
            }

           // ...

            return;
        },
        {
            params: t.Object({ id: t.Integer() }),
            body: t.Object({
                chatID: t.Optional(t.String())}),
        },
    )
```
throwing:
![image](https://github.com/user-attachments/assets/09763628-a63a-47f8-a089-3b67cb84b23e)

**vs**
 
 returning:
![image](https://github.com/user-attachments/assets/ae756ea1-59e9-41ed-a92b-232f5a0a1b4a)


@SaltyAom maybe a good idea to put it into the docs?
